code/game/machinery/doors/firedoor.dm

```diff
@@ -21,14 +22,45 @@
 	assemblytype = /obj/structure/firelock_frame
 	armor = list(MELEE = 10, BULLET = 30, LASER = 20, ENERGY = 20, BOMB = 30, BIO = 100, RAD = 100, FIRE = 95, ACID = 70)
 	interaction_flags_machine = INTERACT_MACHINE_WIRES_IF_OPEN | INTERACT_MACHINE_ALLOW_SILICON | INTERACT_MACHINE_OPEN_SILICON | INTERACT_MACHINE_REQUIRES_SILICON | INTERACT_MACHINE_OPEN
-	var/nextstate = null
 	var/boltslocked = TRUE
+	///List of areas we handle. See CalculateAffectingAreas()
 	var/list/affecting_areas
+	///For the few times we affect only the area we're actually in. Set during Init. If we get moved, we don't update, but this is consistant with fire alarms and also kinda funny so call it intentional.
+	var/area/my_area
+	///Tracks if the firelock is being held open by a crowbar. If so, we don't close until they walk away
 	var/being_held_open = FALSE
+	///Type of alarm when active. See code/defines/firealarm.dm for the list. This var being null means there is no alarm.
+	var/alarm_type = null
+	///Cooldown for Detections. If current world time is not greater than (dectect_cooldown + DETECT_COOLDOWN_STEP_TIME), we don't activate. Prevents instant re-activations when air mixing would solve an issue
+	var/detect_cooldown = 0
+	///The merger_id and merger_typecache variables are used to make rows of firelocks activate at the same time.
+	var/merger_id = "firelocks"
+	var/merger_typecache
+	///Overlay object for the warning lights. This and some plane settings allows the lights to glow in the dark.
+	var/mutable_appearance/warn_lights
 
 /obj/machinery/door/firedoor/Initialize(mapload)
 	. = ..()
+	AddElement(/datum/element/atmos_sensitive, mapload)
 	CalculateAffectingAreas()
+	my_area = get_area(src)
+
+	RegisterSignal(src, COMSIG_MERGER_ADDING, .proc/MergerAdding)
+	RegisterSignal(src, COMSIG_MERGER_REMOVING, .proc/MergerRemoving)
+	if(!merger_typecache)
+		merger_typecache = typecacheof(/obj/machinery/door/firedoor)
+
+	return INITIALIZE_HINT_LATELOAD
+
+/obj/machinery/door/firedoor/LateInitialize()
+	. = ..()
+	GetMergeGroup(merger_id, allowed_types = typecacheof(/obj/machinery/door/firedoor))
```

Comment:
I guess to be more clear.
We generate merger_typecache so it can be used here, because typecacheof is a reasonably expensive proc, and it's best to just use it once.